Facade_Engine/Query/GeneralMaterialTakeoff.cs

```diff
@@ -0,0 +1,177 @@
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Geometry;
+using BH.Engine.Spatial;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using BH.oM.Physical.Constructions;
+using BH.oM.Physical.Elements;
+using BH.oM.Physical.Materials;
+using BH.oM.Facade.Elements;
+using BH.oM.Quantities.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using BH.oM.Facade.SectionProperties;
+
+namespace BH.Engine.Facade
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Gets the general material takeoff from the panel object. The takeoff will contain materials and volumes, mass and areas from the panel itself, as well as takeoffs from any openings.")]
+        [Input("panel", "The facade panel object to extract the general material takeoff from.")]
+        [Output("genTakeoff", "The general material takeoff based on buildup of the panel object as well as any of its openings.")]
+        public static GeneralMaterialTakeoff GeneralMaterialTakeoff(this Panel panel)
+        {
+            if (panel == null)
+            {
+                BH.Engine.Base.Compute.RecordError($"Cannot query the {nameof(GeneralMaterialTakeoff)} of a null {nameof(Panel)}.");
+                return null;
+            }
+
+            //Add null check for opening geo
+            if (panel.ExternalEdges == null)
+            {
+                BH.Engine.Base.Compute.RecordError($"Cannot query the {nameof(GeneralMaterialTakeoff)} of a {nameof(Panel)} with null geometry.");
+                return null;
+            }
+
+            if (panel.Construction == null)
+            {
+                Engine.Base.Compute.RecordError($"The {nameof(GeneralMaterialTakeoff)} could not be queried as no {nameof(IConstruction)} has been assigned to the {nameof(Panel)}.");
+                return null;
+            }
+
+            double openingAreas = 0;
+
+            List<GeneralMaterialTakeoff> takeoffs = new List<GeneralMaterialTakeoff>();
+
+            foreach (Opening opening in panel.Openings)
+            {
+                GeneralMaterialTakeoff openingTakeoff = opening.GeneralMaterialTakeoff();
+                if (openingTakeoff == null)
+                    return null;
```

Comment:
Do you really want to return `null` for the whole panel if one opening returns a null material take off? Seems a bit unnecessary?

Reply:
Fixed, replaced with continue